umbra-js/src/classes/Umbra.ts

```diff
@@ -235,6 +208,81 @@ export class Umbra {
     return { tx, stealthKeyPair };
   }
 
+  async batchSend(signer: JsonRpcSigner | Wallet, sends: SendBatch[], overrides: SendOverrides = {}) {
+    const recipients = new Set(sends.map((send) => send.address));
+    // Check that all recipients are valid
+    await Promise.all([...recipients].map(assertSupportedAddress));
+
+    // Calculate the total amount of each token to be sent
+    const tokenSum = new Map<string, BigNumber>();
+    for (const send of sends) {
+      const token = send.token;
+      const amount = BigNumber.from(send.amount);
+      tokenSum.set(token, tokenSum.get(token)?.add(amount) || amount);
+    }
+
+    // Check that the sender has sufficient balance for all tokens
+    await Promise.all(
+      [...tokenSum.keys()].map((token) => {
+        return assertSufficientBalance(signer, token, tokenSum.get(token) as BigNumber);
+      })
+    );
+
+    const txSigner = this.getConnectedSigner(signer); // signer input validated
+
+    // Get toll amount from contract
+    const toll = await this.umbraContract.toll();
+
+    // Parse provided overrides
+    const { localOverrides, lookupOverrides } = Umbra.parseOverrides({ ...overrides });
+
+    // Prepare data for each send
+    const sendInfo = await Promise.all(
+      sends.map((send) => {
+        return this.prepareSend(send.address, lookupOverrides);
+      })
+    );
+
+    const stealthKeyPairs = sendInfo.map((info) => info.stealthKeyPair);
+    const sendData: SendData[] = [];
+    let valueAmount = BigNumber.from('0');
+
+    sendInfo.map((info, index) => {
+      const receiver = info.stealthKeyPair.address;
+      const token = sends[index].token;
+      const amount = BigNumber.from(sends[index].amount);
+      assertValidStealthAddress(receiver);
+
+      // Sum valueAmount
+      valueAmount = valueAmount.add(toll);
+      if (isEth(token)) {
+        valueAmount = valueAmount.add(amount);
+      }
+
+      sendData.push({
+        receiver: receiver,
+        tokenAddr: token,
+        amount: amount,
+        pkx: info.pubKeyXCoordinate,
+        ciphertext: info.encrypted.ciphertext,
+      });
+    });
```

Comment:
The use case of `map` is to return a new array generated from the old array, typically without side effects. In this case you're not using the return value, and have side effects by calling `sendData.push`.

If you want to keep the side effect approach, using `forEach` instead of `map` would be more semantically correct.

However, a better approach is to directly create and assign `sendData` using the return value from `map`:

```typescript
    const sendData: SendData[] = sendInfo.map((info, index) => {
      const receiver = info.stealthKeyPair.address;
      const token = sends[index].token;
      const amount = BigNumber.from(sends[index].amount);
      assertValidStealthAddress(receiver);

      // Sum valueAmount
      valueAmount = valueAmount.add(toll);
      if (isEth(token)) {
        valueAmount = valueAmount.add(amount);
      }

      return {
        receiver: receiver,
        tokenAddr: token,
        amount: amount,
        pkx: info.pubKeyXCoordinate,
        ciphertext: info.encrypted.ciphertext,
      };
    });
```

Additionally, I don't think we need to iteratively compute `valueAmount`, so that side effect can be removed: our `tokenSum` Map will have a key for the ETH address (or not, in which case start from zero), then sum that value with `toll * sendInfo.length`

Reply:
And actually, why do we need one `.map` to generate `sendInfo` and another one here? Seems those can just be merged into a single `.map` call. And you can generate the `stealthKeyPairs` array afterwards to return it

Reply:
@mds1 So your refactoring comment made sense, but I don't see an obvious way to combine the `sendInfo`' `.map` with `sendData`' `.map` here without losing `Promise.all`, which leads me to something like this:
```
    let stealthKeyPairs: KeyPair[] = [];
    let sendData: SendData[] = [];

    sends.map(async (send) => {
      const { stealthKeyPair, ...info } = await this.prepareSend(send.address, lookupOverrides);
      assertValidStealthAddress(stealthKeyPair.address);

      stealthKeyPairs.push(stealthKeyPair);
      sendData.push({
        receiver: stealthKeyPair.address,
        tokenAddr: send.token,
        amount: BigNumber.from(send.amount),
        pkx: info.pubKeyXCoordinate,
        ciphertext: info.encrypted.ciphertext,
      });
    });
```
Is there a way to combine the two `.maps` without losing `Promise.all`?

Reply:
Does something like this work? (Been a while since I looked at this so a bit hazy on all the discussion here 😅)

```typescript
const sendData = await Promise.all(
  sends.map(async (send) => {
    const sendInfo = await this.prepareSend(send.address, lookupOverrides);
    const receiver = sendInfo.stealthKeyPair.address;
    const token = send.token;
    const amount = BigNumber.from(send.amount);
    assertValidStealthAddress(receiver);

    return {
      receiver: receiver,
      tokenAddr: token,
      amount: amount,
      pkx: sendInfo.pubKeyXCoordinate,
      ciphertext: sendInfo.encrypted.ciphertext,
    };
  })
);
```

Reply:
Ok actually we do not want to do this, because it'd require returning `stealthKeyPair` from this map, but then we'd need to remove that item later to get the right the for the `batchSend` call, so going to mark this resolved